File: VB_Lib/Extensions/Extension_Pix.vb
Imports System.Runtime.InteropServices
Imports System.Drawing
Imports System.Reflection
Imports System.Drawing.Imaging

#Region "Functional"
Partial Public Class Box
    Public Function GetBitmap(Optional ByVal Pix As Pix = Nothing) As Pix
        If IsNothing(Pix) Then Pix = New Pix(Me.w + Me.x + 5, Me.h + Me.y + 5)
        Dim n As New Bitmap(Pix.w, Pix.h, Imaging.PixelFormat.Format24bppRgb)
        Dim g As Graphics = Graphics.FromImage(n)
        Dim p As New Pen(Brushes.Red, 3)
        g.DrawImage(Pix.BitmapStatic, 0, 0, Pix.w, Pix.h)
        g.DrawRectangle(p, New Rectangle(Me.x, Me.y, Me.w, Me.h))
        Dim RPix As New Pix(n) : n.Dispose() : Return RPix
    End Function
    Public Sub Display(Optional Pix As Pix = Nothing)
        Dim n As New ShowPix(GetBitmap(Pix))
    End Sub
End Class

Partial Public Class Boxa
    Public Function GetBitmap(Optional ByVal Pix As Pix = Nothing) As Pix
        Dim MaxW As Integer = 0
        Dim MaxH As Integer = 0
        For Each Entry In Me.box
            If Entry.x + Entry.w + 5 > MaxW Then MaxW = Entry.x + Entry.w + 5
            If Entry.y + Entry.h + 5 > MaxH Then MaxH = Entry.y + Entry.h + 5
        Next
        If IsNothing(Pix) Then Pix = New Pix(MaxW, MaxH)
        Dim n As New Bitmap(Pix.w, Pix.h, Imaging.PixelFormat.Format24bppRgb)
        Dim g As Graphics = Graphics.FromImage(n)
        Dim p As New Pen(Brushes.Red, 3)
        g.DrawImage(Pix.BitmapStatic, 0, 0, Pix.w, Pix.h)
        For Each Boxe In Me.box
            g.DrawRectangle(p, New Rectangle(Boxe.x, Boxe.y, Boxe.w, Boxe.h))
        Next
        Dim RPix As New Pix(n) : n.Dispose() : Return RPix
    End Function
    Public Sub Display(Optional Pix As Pix = Nothing)
        Dim n As New ShowPix(GetBitmap(Pix))
    End Sub
End Class

Partial Public Class Pix
#Region "Redirects"
    ''' <summary>
    ''' This Property returns a cached value of the data array 
    ''' for iterating. 
    ''' </summary>
    ReadOnly Property DataStatic As Byte()
        Get
            If Pointer = IntPtr.Zero Then Return Nothing
            If Caching.ContainsKey("Data1") Then Return Caching("Data1")
            Caching.Add("Data1", Me.data) : Return Caching("Data1")
        End Get
    End Property

    ReadOnly Property BitmapStatic As Bitmap
        Get
            If Pointer = IntPtr.Zero Then Return Nothing
            If Caching.ContainsKey("Bitmap") Then Return Caching("Bitmap")
            Caching.Add("Bitmap", Me.ToBitmap) : Return Caching("Bitmap")
        End Get
    End Property

#Region "Saving"
    Public Sub save_jpg(ByVal filename As String, Optional ByVal quality As Integer = 95, Optional ByVal progressive As Integer = 0)
        LeptonicaSharp.Natives.pixWriteJpeg(filename, Pointer, quality, progressive)
    End Sub
    Public Sub save_png(ByVal filename As String, Optional ByVal gamma As Single = 95, Optional ByVal progressive As Integer = 0)
        LeptonicaSharp.Natives.pixWritePng(filename, Pointer, gamma)
    End Sub
    Public Sub save_tiff(ByVal filename As String)
        LeptonicaSharp.Natives.pixWriteTiff(filename, Pointer, LeptonicaSharp.Enumerations.IFF.IFF_TIFF, "a")
    End Sub
    Public Function save_format(ByVal filename As String, ByVal format As LeptonicaSharp.Enumerations.IFF) As Boolean
        Return LeptonicaSharp.Natives.pixWrite(filename, Pointer, format)
    End Function
    Public Function save_autoformat(ByVal filename As String) As Boolean
        Return LeptonicaSharp.Natives.pixWriteAutoFormat(filename, Pointer)
    End Function
#End Region
#End Region
#Region "Functions"
    Sub RefreshStatic()
        For Each Entry In Caching
            Try : Entry.Value.dispose() : Catch : End Try
        Next : Caching.Clear()
    End Sub
    Sub Display(Optional ByVal Text As String = "")
        Dim n As New ShowPix(Me, Text)
    End Sub
    Public Function ToBitmap() As Image
        Dim Size As Integer = 0
        Dim Bytes As Byte() = Nothing

        If Me.d = 1 Then Return ConvertTo1BPPBMP(Me)
        _All.pixWriteMemBmp(Bytes, Size, IIf(Me.d = 24, Me.ConvertTo32, Me))
        Dim MemStrm As New IO.MemoryStream(Bytes)
        Return New Bitmap(MemStrm, True)
    End Function

    Public Function ConvertTo1BPPBMP(ByVal Pix As Pix) As Bitmap
        Dim MSH As New Marshal_Pix
        Dim PixSwap As Pix = LeptonicaSharp._All.pixEndianByteSwapNew(Pix)

        Try
            Dim img As Bitmap = New Bitmap(Pix.w, Pix.h, PixelFormat.Format1bppIndexed)
            Dim imgData As BitmapData = img.LockBits(New Rectangle(0, 0, img.Width, img.Height), 2, PixelFormat.Format1bppIndexed)
            For Line As Integer = 0 To Pix.h - 1
                For Col = 0 To Pix.w - 1 Step 8
                    Dim index As Integer = (Line * PixSwap.wpl * 4) + (Col >> 3)
                    Marshal.WriteByte(imgData.Scan0, index, Not PixSwap.DataStatic(index))
                Next
            Next
            img.UnlockBits(imgData) : Return img
        Catch

        End Try
        Return Nothing
    End Function
#End Region
End Class

Partial Public Class Pixa
    Function GetBitmap(Optional NCols As Integer = 4) As Pix
        Dim pixc = LeptonicaSharp._All.pixaDisplayTiledAndScaled(Me, 32, 500, NCols, 0, 1, 3)
        Return pixc
    End Function
    Sub Display(Optional NCols As Integer = 4)
        Dim n As New ShowPix(GetBitmap)
    End Sub
End Class

Partial Public Class Numa
    Sub DisplayNumaBarGraph(Optional ByVal NumaH As Numa = Nothing,
                            Optional ByVal BgPix As Pix = Nothing,
                            Optional ByVal Swap As Boolean = False)

        Dim NumaW As Numa = Me
        If Swap = True Then
            Dim TMpH = NumaH
            Dim TmpW = Me
            NumaH = TmpW
            NumaW = TMpH
        End If
        Dim NumaHCnt As Integer = 1 : If Not IsNothing(NumaH) Then NumaHCnt = NumaH.n
        Dim NumaWCnt As Integer = 1 : If Not IsNothing(NumaW) Then NumaWCnt = NumaW.n
        Dim NumaHMax As Double = 1 : If Not IsNothing(NumaH) Then NumaHMax = NumaH.array.Max
        Dim NumaWMax As Double = 1 : If Not IsNothing(NumaW) Then NumaWMax = NumaW.array.Max
        Dim VMaxH As Integer = Math.Max(NumaHMax, NumaWCnt) : If VMaxH > 2000 Then VMaxH = 2000
        Dim VMaxW As Integer = Math.Max(NumaWMax, NumaHCnt) : If VMaxW > 2000 Then VMaxW = 2000
        If Not IsNothing(BgPix) Then VMaxH = BgPix.h : VMaxW = BgPix.w
        Dim DiffH As Double = VMaxH / NumaHMax
        Dim DiffW As Double = VMaxW / NumaWMax

        Dim n As Bitmap = New Bitmap(VMaxW, VMaxH, PixelFormat.Format32bppArgb)
        Dim Pen1 As New Pen(Drawing.Color.FromArgb(150, 0, 255, 0), 1)
        Dim Pen2 As New Pen(Drawing.Color.FromArgb(150, 255, 0, 0), 1)
        Dim g As Graphics = Graphics.FromImage(n)

        If IsNothing(BgPix) Then g.FillRectangle(Brushes.WhiteSmoke, New Rectangle(0, 0, VMaxW, VMaxH))
        If Not IsNothing(BgPix) Then g.DrawImage(BgPix.ToBitmap, New System.Drawing.Point(0, 0))

        If Not IsNothing(NumaW) Then
            For i As Integer = 0 To NumaW.n - 1
                g.DrawLine(Pen1, New Drawing.Point(0, i + 1), New Drawing.Point(NumaW.array(i) * DiffW, i + 1))
            Next
        End If
        If Not IsNothing(NumaH) Then
            For i As Integer = 0 To NumaH.n - 1
                g.DrawLine(Pen2, New Drawing.Point(i + 1, VMaxH - (NumaH.array(i) * DiffH)), New Drawing.Point(i + 1, VMaxH))
            Next
        End If
        Dim PX As New Pix(n)
        PX.Display()
        PX.Dispose()
    End Sub
    Sub DisplayNumaHeatmap(Optional ByVal NumaH As Numa = Nothing,
                           Optional ByVal BgPix As Pix = Nothing,
                           Optional ByVal Swap As Boolean = False)

        Dim NumaW As Numa = Me
        If Swap = True Then
            Dim TMpH = NumaH
            Dim TmpW = NumaW
            NumaH = TmpW
            NumaW = TMpH
        End If
        Dim NumaHCnt As Integer = 1 : If Not IsNothing(NumaH) Then NumaHCnt = NumaH.n
        Dim NumaWCnt As Integer = 1 : If Not IsNothing(NumaW) Then NumaWCnt = NumaW.n
        Dim NumaHMax As Double = 1 : If Not IsNothing(NumaH) Then NumaHMax = NumaH.array.Max
        Dim NumaWMax As Double = 1 : If Not IsNothing(NumaW) Then NumaWMax = NumaW.array.Max
        Dim VMaxH As Integer = Math.Max(NumaHMax, NumaWCnt) : If VMaxH > 2000 Then VMaxH = 2000
        Dim VMaxW As Integer = Math.Max(NumaWMax, NumaHCnt) : If VMaxW > 2000 Then VMaxW = 2000
        If Not IsNothing(BgPix) Then VMaxH = BgPix.h : VMaxW = BgPix.w
        Dim VDiffH As Double = VMaxH / NumaHMax
        Dim VDiffW As Double = VMaxW / NumaWMax

        Dim n As Bitmap = New Bitmap(VMaxW, VMaxH, PixelFormat.Format32bppArgb)
        Dim g As Graphics = Graphics.FromImage(n)

        If IsNothing(BgPix) Then g.FillRectangle(Brushes.WhiteSmoke, New Rectangle(0, 0, VMaxW, VMaxH))
        If Not IsNothing(BgPix) Then g.DrawImage(BgPix.ToBitmap, New System.Drawing.Point(0, 0))

        If Not IsNothing(NumaW) Then
            Dim BarHeight As Integer = CInt(VMaxH / NumaWCnt)
            For i As Integer = 0 To NumaW.n - 1
                Dim BarPos As Integer = ((i) * BarHeight) + IIf(BarHeight = 1, 0, (BarHeight / 2))
                Dim Pen As New Pen(Color.FromArgb(CInt((130 / NumaWMax) * NumaW.array(i)), 0, 255, 0), BarHeight)
                Dim P1 As New Drawing.Point(0, BarPos)
                Dim P2 As New Drawing.Point(VMaxW, BarPos)
                g.DrawLine(Pen, P1, P2)
            Next
        End If

        If Not IsNothing(NumaH) Then
            Dim BarWidth As Integer = Math.Max(CInt(VMaxW / NumaHCnt), 1)
            For i As Integer = 0 To NumaH.n - 1
                Dim BarPos As Integer = ((i) * BarWidth) + IIf(BarWidth = 1, 0, (BarWidth / 2))
                Dim Pen As New Pen(Color.FromArgb(CInt((130 / NumaHMax) * NumaH.array(i)), 0, 0, 255), BarWidth)
                Dim P1 As New Drawing.Point(BarPos, 0)
                Dim P2 As New Drawing.Point(BarPos, VMaxH)
                g.DrawLine(Pen, P1, P2)
            Next
        End If

        Dim PX As New Pix(n)
        PX.Display()
        PX.Dispose()
    End Sub
End Class

Partial Public Class Sel
    Sub Display()
        Dim n As New ShowPix(LeptonicaSharp._All.selDisplayInPix(Me, 1, 1))
    End Sub
    Sub Display(ByVal Pix As Pix)
        Dim CRed As UInt32 = BitConverter.ToUInt32({255, 0, 0, 0}, 0)
        Dim CBlu As UInt32 = BitConverter.ToUInt32({0, 255, 0, 0}, 0)
        Dim n As New ShowPix(LeptonicaSharp._All.pixDisplayHitMissSel(Pix, Me, 1, CRed, CBlu))
    End Sub
End Class
#End Region

File: CS_LIB_ALPHA/Extensions/Extension_Pix.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Drawing;
using System.Reflection;
using System.Drawing.Imaging;
using LeptonicaSharp;
using LeptonicaCSharp.Extensions;

    public partial class Box
    {
        public Pix GetBitmap(Pix Pix = null)
        {
            if ((Pix == null))
                Pix = new Pix(this.w + this.x + 5, this.h + this.y + 5);
            Bitmap n = new Bitmap((int) Pix.w,(int) Pix.h, System.Drawing.Imaging.PixelFormat.Format24bppRgb);
            Graphics g = Graphics.FromImage(n);
            Pen p = new Pen(Brushes.Red, 3);
            g.DrawImage(Pix.BitmapStatic, 0, 0, Pix.w, Pix.h);
            g.DrawRectangle(p, new Rectangle(this.x, this.y, this.w, this.h));
            Pix RPix = new Pix(n);
            n.Dispose();
            return RPix;
        }
        public void Display(Pix Pix = null)
        {
            ShowPix n = new ShowPix(GetBitmap(Pix));
        }
    }

    public partial class Boxa
    {
        public Pix GetBitmap(Pix Pix = null)
        {
            int MaxW = 0;
            int MaxH = 0;
            foreach (Box Entry_loopVariable in this.box)
            {
               Box Entry = Entry_loopVariable;
                if (Entry.x + Entry.w + 5 > MaxW)
                    MaxW = Entry.x + Entry.w + 5;
                if (Entry.y + Entry.h + 5 > MaxH)
                    MaxH = Entry.y + Entry.h + 5;
            }
            if ((Pix == null))
                Pix = new Pix(MaxW, MaxH);
            Bitmap n = new Bitmap((int) Pix.w, (int)Pix.h, System.Drawing.Imaging.PixelFormat.Format24bppRgb);
            Graphics g = Graphics.FromImage(n);
            Pen p = new Pen(Brushes.Red, 3);
            g.DrawImage(Pix.BitmapStatic, 0, 0, Pix.w, Pix.h);
            foreach (Box Boxe_loopVariable in this.box)
            {
               Box Boxe = Boxe_loopVariable;
                g.DrawRectangle(p, new Rectangle(Boxe.x, Boxe.y, Boxe.w, Boxe.h));
            }
            Pix RPix = new Pix(n);
            n.Dispose();
            return RPix;
        }
        public void Display(Pix Pix = null)
        {
            ShowPix n = new ShowPix(GetBitmap(Pix));
        }
    }

    public partial class Pix
    {
        #region "Redirects"
        /// <summary>
        /// This Property returns a cached value of the data array 
        /// for iterating. 
        /// </summary>
        public byte[] DataStatic
        {
            get
            {
                if (Pointer == IntPtr.Zero)
                    return null;
                if (Caching.ContainsKey("Data1"))
                    return (byte[]) Caching["Data1"];
                Caching.Add("Data1", this.data);
                return (byte[]) Caching["Data1"];
            }
        }

        public Bitmap BitmapStatic
        {
            get
            {
                if (Pointer == IntPtr.Zero)
                    return null;
                //if (Caching.ContainsKey("Bitmap")) {CType (Caching ("Bitmap"), System.Drawing.Bitmap ). , Caching.Remove("Bitmap"); }
                Caching.Add("Bitmap", this.ToBitmap());
                return (Bitmap ) Caching["Bitmap"];
            }
        }

        #region "Saving"
        public void save_jpg(string filename, int quality = 95, int progressive = 0)
        {
            Natives.pixWriteJpeg(filename, Pointer, quality, progressive);
        }
        public void save_png(string filename, float gamma = 95, int progressive = 0)
        {
            Natives.pixWritePng(filename, Pointer, gamma);
        }
        public void save_tiff(string filename)
        {
            Natives.pixWriteTiff(filename,Pointer, (int) Enumerations.IFF.IFF_TIFF , "a");
        }
        public int save_format(string filename, Enumerations.IFF format)
        {
            return Natives.pixWrite(filename, Pointer,(int) format);
        }
        public int save_autoformat(string filename)
        {
            return Natives.pixWriteAutoFormat(filename,Pointer);
        }
        #endregion
        #endregion
        #region "Functions"
        public void RefreshStatic()
        {
            foreach (KeyValuePair<string, object> Entry in Caching)
            {
                try
                {
                    //Entry.Value = null;
                }
                catch
                {
                }
            }
            Caching.Clear();
        }
        public void Display(string Text = "")
        {
            ShowPix n = new ShowPix(this, Text);
        }
        public Image ToBitmap()
        {
            uint Size = 0;
            byte[] Bytes = null;

            if (this.d == 1) { return ConvertTo1BPPBMP(this); }
            _All.pixWriteMemBmp(out Bytes, out Size, (this.d == 24 ? this.pixConvertTo32() : this));
            System.IO.MemoryStream MemStrm = new System.IO.MemoryStream(Bytes);
            return new Bitmap(MemStrm, true);
        }

        public Bitmap ConvertTo1BPPBMP(Pix Pix)
        {
            Marshal_Pix MSH = new Marshal_Pix();
            Pix PixSwap = LeptonicaSharp._All.pixEndianByteSwapNew(Pix);

            try
            {
                Bitmap img = new Bitmap((int) Pix.w, (int)Pix.h, PixelFormat.Format1bppIndexed );
                BitmapData imgData = img.LockBits(new Rectangle(0, 0, (int)img.Width, (int)img.Height), ImageLockMode.ReadOnly  , PixelFormat.Format1bppIndexed);
                for (int Line = 0; Line <= Pix.h - 1; Line++)
                {
                    for (int Col = 0; Col <= Pix.w - 1; Col += 8)
                    {
                        int index = ((int)Line * (int)PixSwap.wpl * 4) + (Col >> 3);
                        Byte Byt =  PixSwap.DataStatic[index];
                        Marshal.WriteByte(imgData.Scan0, index, (Byte) (Byt ^ 0xff));
                    }
                }
                img.UnlockBits(imgData);
                return img;
                }
            catch
            {
            }
            return null;
        }
        #endregion
    }

    public partial class Pixa
    {
        public Pix GetBitmap(int NCols = 4)
        {
            var pixc = LeptonicaSharp._All.pixaDisplayTiledAndScaled(this, 32, 500, NCols, 0, 1, 3);
            return pixc;
        }
        public void Display(int NCols = 4)
        {
            ShowPix n = new ShowPix(GetBitmap());
        }
    }

    public partial class Numa
    {
        public void DisplayNumaBarGraph(Numa NumaH = null, Pix BgPix = null, bool Swap = false)
        {
        //    Numa NumaW = this;
        //    if (Swap == true)
        //    {
        //        var TMpH = NumaH;
        //        var TmpW = this;
        //        NumaH = TmpW;
        //        NumaW = TMpH;
        //    }
        //    int NumaHCnt = 1;
        //    if ((NumaH != null))
        //        NumaHCnt = NumaH.n;
        //    int NumaWCnt = 1;
        //    if ((NumaW != null))
        //        NumaWCnt = NumaW.n;
        //    //double NumaHMax = 1;
        //    //if ((NumaH != null))
        //    //    NumaHMax = (int) NumaH.array.Max();
        //    //double NumaWMax = 1;
        //    //if ((NumaW != null))
        //    //    NumaWMax = (int) NumaW.array.Max();
        //    //int VMaxH = (int )Math.Max(NumaHMax, NumaWCnt);
        //    //if (VMaxH > 2000)
        //    //    VMaxH = 2000;
        //    //int VMaxW = (int) Math.Max(NumaWMax, NumaHCnt);
        //    //if (VMaxW > 2000)
        //    //    VMaxW = 2000;
        //    //if ((BgPix != null)) { VMaxH = (int) BgPix.h; VMaxW = (int) BgPix.w; }
        //    //double DiffH = VMaxH / NumaHMax;
        //    //double DiffW = VMaxW / NumaWMax;

        //    //Bitmap n = new Bitmap(VMaxW, VMaxH, PixelFormat.Format32bppArgb);
        //    Pen Pen1 = new Pen(System.Drawing.Color.FromArgb(150, 0, 255, 0), 1);
        //    Pen Pen2 = new Pen(System.Drawing.Color.FromArgb(150, 255, 0, 0), 1);
        //    //Graphics g = Graphics.FromImage(n);

        //    if ((BgPix == null))
        //        //g.FillRectangle(Brushes.WhiteSmoke, new Rectangle(0, 0, VMaxW, VMaxH));
        //    if ((BgPix != null))
        //        //g.DrawImage(BgPix.ToBitmap(), new System.Drawing.Point(0, 0));

        //    if ((NumaW != null))
        //    {
        //        for (int i = 0; i <= NumaW.n - 1; i++)
        //        {
        //            g.DrawLine(Pen1, new System.Drawing.Point(0, i + 1), new System.Drawing.Point((int) NumaW.array[i] * (int) DiffW, i + 1));
        //        }
        //    }
        //    if ((NumaH != null))
        //    {
        //        for (int i = 0; i <= NumaH.n - 1; i++)
        //        {
        //            g.DrawLine(Pen2, new System.Drawing.Point(i + 1, (int) VMaxH - ((int)NumaH.array[i] * (int)DiffH)), new System.Drawing.Point(i + 1, VMaxH));
        //        }
        //    }
        //    Pix PX = new Pix(n);
        //    PX.Display();
        //    PX.Dispose();
        }
        public void DisplayNumaHeatmap(Numa NumaH = null, Pix BgPix = null, bool Swap = false)
        {
        //    Numa NumaW = this;
        //    if (Swap == true)
        //    {
        //        var TMpH = NumaH;
        //        var TmpW = NumaW;
        //        NumaH = TmpW;
        //        NumaW = TMpH;
        //    }
        //    int NumaHCnt = 1;
        //    if ((NumaH != null))
        //        NumaHCnt = NumaH.n;
        //    int NumaWCnt = 1;
        //    if ((NumaW != null))
        //        NumaWCnt = NumaW.n;
        //    double NumaHMax = 1;
        //    if ((NumaH != null))
        //        NumaHMax = (int)NumaH.array.Max();
        //    double NumaWMax = 1;
        //    if ((NumaW != null))
        //        NumaWMax = (int)NumaW.array.Max();
        //    int VMaxH = (int)(Math.Max(NumaHMax, NumaWCnt));
        //    if (VMaxH > 2000)
        //        VMaxH = 2000;
        //    int VMaxW = (int)(Math.Max(NumaWMax, NumaHCnt));
        //    if (VMaxW > 2000)
        //        VMaxW = 2000;
        //    if ((BgPix != null)) { VMaxH = (int) BgPix.h; VMaxW = (int)BgPix.w; }
        //    double VDiffH = VMaxH / NumaHMax;
        //    double VDiffW = VMaxW / NumaWMax;

        //    Bitmap n = new Bitmap(VMaxW, VMaxH, PixelFormat.Format32bppArgb);
        //    Graphics g = Graphics.FromImage(n);

        //    if ((BgPix == null))
        //        g.FillRectangle(Brushes.WhiteSmoke, new Rectangle(0, 0, VMaxW, VMaxH));
        //    if ((BgPix != null))
        //        g.DrawImage(BgPix.ToBitmap(), new System.Drawing.Point(0, 0));

        //    if ((NumaW != null))
        //    {
        //        int BarHeight = System.Convert.ToInt32 (VMaxH / NumaWCnt);
        //        for (int i = 0; i <= NumaW.n - 1; i++)
        //        {
        //            int BarPos = ((i) * BarHeight) + (BarHeight == 1 ? 0 : (BarHeight / 2));
        //            Pen Pen = new Pen(Color.FromArgb(System.Convert.ToInt32((130 / NumaWMax) * NumaW.array[i]), 0, 255, 0), BarHeight);
        //            System.Drawing.Point P1 = new System.Drawing.Point(0, BarPos);
        //            System.Drawing.Point P2 = new System.Drawing.Point(VMaxW, BarPos);
        //            g.DrawLine(Pen, P1, P2);
        //        }
        //    }

        //    if ((NumaH != null))
        //    {
        //        int BarWidth = Math.Max(System.Convert.ToInt32(VMaxW / NumaHCnt), 1);
        //        for (int i = 0; i <= NumaH.n - 1; i++)
        //        {
        //            int BarPos = ((i) * BarWidth) + (BarWidth == 1 ? 0 : (BarWidth / 2));
        //            Pen Pen = new Pen(Color.FromArgb(System.Convert.ToInt32((130 / NumaHMax) * NumaH.array[i]), 0, 0, 255), BarWidth);
        //            System.Drawing.Point P1 = new System.Drawing.Point(BarPos, 0);
        //            System.Drawing.Point P2 = new System.Drawing.Point(BarPos, VMaxH);
        //            g.DrawLine(Pen, P1, P2);
        //        }
        //    }

        //    Pix PX = new Pix(n);
        //    PX.Display();
        //    PX.Dispose();
        }
    }

    public partial class Sel
    {
        public void Display()
        {
           ShowPix n = new ShowPix(LeptonicaSharp._All.selDisplayInPix(this, 1, 1));
        }
        public void Display(Pix Pix)
		{
			UInt32 CRed = BitConverter.ToUInt32( new byte[]{255,0,0,0}, 0);
            UInt32 CBlu = BitConverter.ToUInt32(new byte[]{0,255,0,0}, 0);
			ShowPix n = new ShowPix(LeptonicaSharp._All.pixDisplayHitMissSel(Pix, this, 1, CRed, CBlu));
		}
    }
